python/pyarrow/tests/test_convert_builtin.py

```diff
@@ -334,14 +362,77 @@ def test_garbage_collection():
 
 
 def test_sequence_double():
-    data = [1.5, 1, None, 2.5, None, None]
+    data = [1.5, 1., None, 2.5, None, None]
     arr = pa.array(data)
     assert len(arr) == 6
     assert arr.null_count == 3
     assert arr.type == pa.float64()
     assert arr.to_pylist() == data
 
 
+def test_double_auto_coerce_from_integer():
+    # Done as part of ARROW-2814
+    data = [1.5, 1., None, 2.5, None, None]
+    arr = pa.array(data)
+
+    data2 = [1.5, 1, None, 2.5, None, None]
+    arr2 = pa.array(data2)
+
+    assert arr.equals(arr2)
+
+    data3 = [1, 1.5, None, 2.5, None, None]
+    arr3 = pa.array(data3)
+
+    data4 = [1., 1.5, None, 2.5, None, None]
+    arr4 = pa.array(data4)
+
+    assert arr3.equals(arr4)
+
+
+def test_double_integer_coerce_representable_range():
+    valid_values = [1.5, 1, 2, None, 1 << 53, -(1 << 53)]
+    invalid_values = [1.5, 1, 2, None, 1 << 53 + 1]
```

Comment:
Python's operator precedence is a bit surprising here: `1 << 53 + 1` is equal to `1 << 54`.

(same below with negative values and floats)

Reply:
Indeed I found this surprising. fixing